### PR TITLE
Handle unhandled Promise rejections in JavaScript

### LINKED ISSUE #62: `panic` does not trigger a failed test when within a JavaScript Promise
I was experimenting with tests involving large amounts of JS Promises, and noticed that gleeunit ran into some issues with code involving Promises.

Here is a minimal example that passes in gleeunit but fails when run directly:

```gleam
import gleam/javascript/promise
import gleeunit

pub fn main() -> Nil {
  gleeunit.main()
}

pub fn main_test() -> Nil {
  echo "testing"
  let promise1 =
    promise.new(fn(resolve) {
      echo "testing promise"
      panic
      resolve("Hello, World!")
    })

  Nil
}
```

The expected behavior is to match Node/Deno/Bun, which all throw errors in execution.

### PR DESCRIPTION
**Pull request by acandoo**
Fixes #62

# About

On all major server-side JavaScript runtimes, unhandled rejections of Promises by default lead to a fatal error and the program prematurely exiting. This can be demonstrated in pure Gleam:

```gleam
//// test/wibble_test.gleam

import gleam/javascript/promise
import gleeunit

pub fn main() -> Nil {
  // All tests pass!
  gleeunit.main()
}

// This fails when run directly!
pub fn main_test() -> Nil {
  echo "testing"
  let promise1 =
    promise.new(fn(resolve) {
      echo "testing promise"
      panic
      resolve("Hello, World!")
    })

  Nil
}
```

However, attaching an event listener to unhandled rejections changes this behavior, and allows for detecting these errors. This PR handles these rejections and appends it as a failed test case from an unknown module and unknown function.

For testing, this PR also adds two dev dependencies `envoy` and `shellout`, as testing requires the test runner explicitly failing.

# Caveats

This code assumes that ALL unhandled rejections within tests should be considered as failures. In some programs this may not be the case, as unhandled rejections don't ALWAYS lead to program exit if the program also subscribes to unhandled rejections through `process.on`. Some Gleam programs may want this behavior, but as it would lead to `assert`s and `panic`s within unhandled rejections not being recognized. This code could be changed to either:
- mark all unhandled rejections as failed tests **(current behavior)**
- ignore unhandled rejections if they're already handled by the program
- only mark Gleam-specific errors as test failures
- some combination of the previous two

### CODE REVIEW

**Comment by acandoo** [comment]
oh gawd this is a rabbit hole...

so catching unhandled rejections through `process.on` effectively catches the promise, but doing it through `globalThis.addEventListener` or `globalThis.onunhandledrejection` just runs an event. you can "catch" the promise through either of the two and it will trigger the "rejectionhandled" event, but the process still exits with an error.

**Comment by acandoo** [comment]
okay wait i think this makes life easier, i just have to check for nodejs-style catching of unhandled rejections, force push coming soon

**Comment by acandoo** [comment]
ok updated!

while i was at it i added logic so that process exit occurs at `'beforeExit'`, ie when the runtime has nothing else (like timers) keeping the program alive.

i also now made it so that all unhandled rejections are assumed to be failures, even if the program itself subscribes to unhandled rejections or handles it at a point where the program would normally crash without subscribing. i figured that logic wasn't worth keeping since realistically promise rejection should only happen when you WANT an exception to occur in the test. though i'll update the description because both behaviors to me make sense.

File: test/gleam_panics_test.gleam [hunk 1]
```diff
@@ -206,3 +208,65 @@ pub fn assert_binary_operator_test() {
...
+    Ok("1") -> {
+      echo "inside spawned gleam test, testing immediate reject"
+      javascript_test_promise()
+    }
+    Ok(_) -> {
+      echo "inside spawned gleam test, testing delayed reject"
+      javascript_test_delayed_promise()
+    }
+  }
+}
```
Comment (lpil) [inline test/gleam_panics_test.gleam:258]: I don't understand this test, what is it doing? Why is it reading environment variables and spawning OS processes?

File: test/gleeunit_test_ffi.mjs [hunk 2]
```diff
@@ -1,9 +1,28 @@
 import { Ok, Error } from "./gleam.mjs";
 
+// The maximum amount of time running all tests is expected to take.
+// This is used to set the timeout for the delayed promise test, to ensure
+// it runs after all other tests have had a chance to execute.
+const MAX_TEST_TIME = 4000;
```
Comment (lpil) [inline test/gleeunit_test_ffi.mjs:6]: This comment is a bit confusing to me, it sounds like it is being used as the timeout for tests being run, but it seems to instead be used to reject one promise in the FFI below?

File: test/gleeunit_test_ffi.mjs [hunk 3]
```diff
@@ -1,9 +1,28 @@
...
+export function promise_fail_test() {
+  new Promise(() => {
+    throw new Error("Promise panicked");
+  });
+}
+
+export function delayed_promise_fail_test() {
+  new Promise((_, reject) => {
+    setTimeout(() => {
+      reject(new Error("Promise panicked"));
```
Comment (lpil) [inline test/gleeunit_test_ffi.mjs:25]: This says the promise panics, but there's no panic here.

File: src/gleeunit_ffi.mjs [hunk 4]
```diff
@@ -55,8 +70,28 @@ export async function main() {
...
+    for (const entry of rejections) {
+      state = reporting.test_failed(state, "Unknown module", "Unknown function", entry);
+    }
+
+    const status = reporting.finished(state);
+    exit(status);
+  };
+
+  // If we're able to subscribe to beforeExit, then we can report unhandled rejections that
+  // occur after the tests are finished.
```
Comment (lpil) [inline src/gleeunit_ffi.mjs:89]: Very cool! Which runtimes is this API not available for?

**Review by lpil — COMMENTED** [review]
Thank you! Fantastic work! I've left some questions inline.

File: test/gleeunit_test_ffi.mjs (hunk 2 above)
Comment (acandoo) [reply test/gleeunit_test_ffi.mjs:6]: The intention is for the Promise's rejection to outlast the regular time of the tests, as it's testing to make sure `gleeunit` doesn't prematurely exit before the Promise rejects.

File: test/gleam_panics_test.gleam (hunk 1 above)
Comment (acandoo) [reply test/gleam_panics_test.gleam:258]: It respawns `gleam test` but using the JavaScript runtime and with an environment variable, as the test in question requires seeing if `gleeunit` exits with an error on an uncaught Promise rejection. Since there are two tests (immediate unhandled rejection, delayed unhandled rejection), it's spawned two times with the environment variable set for the two test cases to run. I was considering either this or making an OS temporary directory with a brand-new Gleam project, but figured this was easier to implement.

File: src/gleeunit_ffi.mjs (hunk 4 above)
Comment (acandoo) [reply src/gleeunit_ffi.mjs:89]: Node, Deno, and Bun all cover it, so the edge case is if someone decides to run it on a Vercel Edge runtime or some really early version of Deno/Bun or a browser or something :p i can remove it if you want

File: test/gleeunit_test_ffi.mjs (hunk 3 above)
Comment (acandoo) [reply test/gleeunit_test_ffi.mjs:25]: Got it, i'll update the message!

File: test/gleam_panics_test.gleam (hunk 1 above)
Comment (lpil) [reply test/gleam_panics_test.gleam:258]: I see! Could you replace this with a new package that have test failures and then snapshot with birdie the output of running `gleam test` in that project 🙏 Thank you.

File: test/gleeunit_test_ffi.mjs [hunk 5]
```diff
@@ -1,9 +1,28 @@
...
+
 export function rescue(f) {
   try {
     return new Ok(f());
   } catch (e) {
     return new Error(e);
   }
 }
+
+export function promise_fail_test() {
```
Comment (lpil) [inline test/gleeunit_test_ffi.mjs:16]: In the new package use `gleam_javascript` instead of FFI for these please.

**Review by lpil — COMMENTED** [review]
Thank you! I've left some notes inline
